platform/darwin/src/MGLMapSnapshotter.h

```diff
@@ -284,6 +288,78 @@ MGL_EXPORT
+ snapshotting process, such as the style loading. Implement a delegate to
+ customize the style that is depicted by the snapshot.
+ 
+ You set the delegate after initializing the snapshotter but before receiving
+ the snapshot, typically before starting the snapshot. The snapshotter keeps a
+ weak reference to its delegate, so you must keep a strong reference to it to
+ ensure that your style customizations apply.
+ */
+@property (nonatomic, weak) id <MGLMapSnapshotterDelegate> delegate;
+
+/**
+ The style displayed in the resulting snapshot.
+ 
+ Unlike the `MGLMapSnapshotOptions.styleURL` property, this property is set to
+ an object that allows you to manipulate every aspect of the style locally.
+ 
+ This property is set to `nil` until the style finishes loading. If the style
+ has failed to load, this property is set to `nil`. Because the style loads
+ asynchronously, you should manipulate it in the
+ `-[MGLMapSnapshotterDelegate mapSnapshotter:didFinishLoadingStyle:]` method. It
+ is not possible to manipulate the style before it has finished loading.
+
+ @note The default styles provided by Mapbox contain sources and layers with
+    identifiers that will change over time. Applications that use APIs that
+    manipulate a style’s sources and layers must first set the style URL to an
+    explicitly versioned style using a convenience method like
+    `+[MGLStyle outdoorsStyleURLWithVersion:]` or a manually constructed
+    `NSURL`.
+ */
+@property (nonatomic, readonly, nullable) MGLStyle *style;
+
+@end
+
+/**
+ Optional methods about significant events when creating a snapshot using an
+ `MGLMapSnapshotter` object.
+ */
+@protocol MGLMapSnapshotterDelegate <NSObject>
+@optional
+
+/**
+ Tells the delegate that the snapshotter was unable to load data needed for
+ snapshotting the map.
+ 
+ This method may be called for a variety of reasons, including a network
+ connection failure or a failure to fetch the style from the server. You can use
+ the given error message to notify the user that map data is unavailable.
+ 
+ @param snapshotter The snapshotter that is unable to load the data.
+ @param error The reason the data could not be loaded.
+*/
+- (void)mapSnapshotterDidFail:(MGLMapSnapshotter *)snapshotter withError:(NSError *)error;
+
+/**
+ Tells the delegate that the snapshotter has just finished loading a style.
+ 
+ This method is called in response to
+ `-[MGLMapSnapshotter startWithQueue:completionHandler:]` as long as the
+ `MGLMapSnapshotter.delegate` property is set. Changes to sources or layers of
+ the style being snapshotted do not cause this method to be called.
+ 
+ @param snapshotter The snapshotter that has just loaded a style.
+ @param style The style that was loaded.
+ */
+- (void)mapSnapshotter:(MGLMapSnapshotter *)snapshotter didFinishLoadingStyle:(MGLStyle *)style;
+
+- (void)mapSnapshotter:(MGLMapSnapshotter *)snapshotter didFailLoadingImageNamed:(NSString *)name;
```

Comment:
If we choose not to support this delegate method, should we move the `mapSnapshotter:didFinishLoadingStyle:` to another block handler? Alternatively, if we feel using a delegate is preferable how about we add a finish method and deprecate the handlers? (I think you discussed this before, so apologies if I'm just catching up).

Reply:
I went back and forth on whether to introduce an additional block parameter as a convenience to avoid having to write a delegate. I would’ve liked to use the “action block” pattern in https://github.com/mapbox/mapbox-gl-native-ios/issues/200#issuecomment-598464934. The problem is that there are already two block parameters that get called on different queues, so adding a third results in a pretty complicated call site plus confusion over which blocks get called on which queues.

Reply:
👍 Let's consider deprecating the overlay handler in favour of a delegate method in a future release?

Reply:
The difference is that the delegate methods are all called on the same thread that the snapshot was started on. Typically this would be the main thread. (This is how mapbox/mapbox-gl-native#16268 was designed.) Based on the discussion in https://github.com/mapbox/mapbox-gl-native-ios/pull/210#discussion_r392107428, it’s unlikely that we’d move the overlay handler to the main thread. I’m wary of encouraging the developer to write a class whose methods are called on different threads, which would require internal synchronization that even the best developers get wrong sometimes.